The job **`x86_64-gnu-aux`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15798166815/job/44532956248) [(plain)](https://github.com/rust-lang/rust/commit/e28da9c9042374f1b7afd4520e9ff0562026ddbc/checks/44532956248/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
warning: `iron-exmaples` (bin "around" test) generated 3 warnings (run `cargo fix --bin "around" --tests` to apply 3 suggestions)
warning: `iron` (lib test) generated 44 warnings (37 duplicates) (run `cargo fix --lib -p iron --tests` to apply 5 suggestions)
error: linking with `cc` failed: exit status: 1
  |
  = note:  "cc" "-m64" "/tmp/rustcpopvuu/symbols.o" "<203 object files omitted>" "-Wl,--as-needed" "-Wl,-Bstatic" "/checkout/obj/build/ct/iron/target/debug/deps/{libmount-6e9c66b312103458.rlib,libsequence_trie-5a35fa00d54735d5.rlib,libiron-fc0d0ae2b658f0c7.rlib,libmodifier-4fb1f7e18ad72722.rlib,liburl-2215c3d75c1b8584.rlib,libpercent_encoding-c749af53d9e74952.rlib,libidna-e7e6c70386fa9498.rlib,libunicode_normalization-7e7d0af3110d8833.rlib,libtinyvec-ff29d6288924b12a.rlib,libtinyvec_macros-9107a8cc40680900.rlib,libunicode_bidi-696686a2806343bd.rlib,libmatches-46d1528f77d431eb.rlib,libplugin-f3d0c71b95e43ed1.rlib,libtypemap-093486427740f16d.rlib,libunsafe_any-4930bbaaf7015225.rlib,libtraitobject-340c08327ba62746.rlib,libmime_guess-69adc7798ff3a188.rlib,libunicase-4060a32050b12bb2.rlib,libmime-331a3ecb015ebe30.rlib,libhyper-4cb0a79202e50e61.rlib,libwant-6f13659271f508f4.rlib,libtry_lock-3d6719b540eeb8b5.rlib,libtokio_tcp-d5b94f13d1b96cb6.rlib,libtokio-de0d43b6b561a441.rlib,libtokio_timer-77f3ccedc641dfef.rlib,libtokio_threadpool-6ed6059afa757ecf.rlib,libcrossbeam_queue-18a1a169f95c2544.rlib,libcrossbeam_deque-52bc3a222f80faea.rlib,libcrossbeam_epoch-0b3cda5c0f51574c.rlib,libmemoffset-8dc40f9b9d911ae1.rlib,libtokio_reactor-42930f800d526309.rlib,libtokio_sync-4ad007f9322fa3e5.rlib,libparking_lot-217325a0d93a6aef.rlib,libparking_lot_core-538ce51d58966a81.rlib,libsmallvec-646270aae7e48443.rlib,libmaybe_uninit-dcfcb0c426b7fe3f.rlib,liblock_api-b71f254baa560a48.rlib,libscopeguard-fb83805fc30b1764.rlib,libtokio_current_thread-f8b4e415ea556471.rlib,libtokio_executor-0dbedb92b823b9f2.rlib,libcrossbeam_utils-8d5f9c2a22510913.rlib,liblazy_static-685cca6a61de9387.rlib,libmio-13e00c0d505d782f.rlib,libtime-fddd10bab50ff5b2.rlib,libnet2-fb030a73ce79c7da.rlib,libcfg_if-399308ddf934bdc9.rlib,libhttparse-96ab9397b6935093.rlib,libhttp_body-c9cb5fb9994bb288.rlib,libtokio_buf-df7e18f9dd7750e0.rlib,libh2-7f455a0f03742b8f.rlib,libindexmap-821677bb675840bc.rlib,libhashbrown-6ea7667305d8e2be.rlib,libstring-4b7f892e145a84c8.rlib,libslab-33d2b2db3b3a1c98.rlib,libtokio_io-d3ec67a4e3810a84.rlib,libhttp-bba1012344b91bed.rlib,libitoa-1c0dfcba93f6c2bd.rlib,libfnv-01d75fecb65da78b.rlib,libbytes-62948dbb652bab20.rlib,libeither-f888497e747bd327.rlib,libiovec-4a836b323fd99988.rlib,libbyteorder-df09cad6ecc0a915.rlib,libfutures_cpupool-cb467489621a9287.rlib,libnum_cpus-123ade2ff149fee0.rlib,liblibc-c1672baac6be931e.rlib,libfutures-674fdf4396be79cd.rlib,liblog-3c76e791db33a3b7.rlib}.rlib" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib/{libstd-*,libpanic_unwind-*,libobject-*,libmemchr-*,libaddr2line-*,libgimli-*,librustc_demangle-*,libstd_detect-*,libhashbrown-*,librustc_std_workspace_alloc-*,libminiz_oxide-*,libadler2-*,libunwind-*,libcfg_if-*,liblibc-*,librustc_std_workspace_core-*,liballoc-*,libcore-*,libcompiler_builtins-*}.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustcpopvuu/raw-dylibs" "-B<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-fuse-ld=lld" "-Wl,-znostart-stop-gc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/ct/iron/target/debug/examples/double_mount-e56c926f98951294" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: rust-lld: error: undefined symbol: core::ptr::drop_in_place$LT$usize$GT$::h55c321e72c91cfba
          >>> referenced by mod.rs:686 (/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hashbrown-0.12.3/src/raw/mod.rs:686)
          >>>               h2-7f455a0f03742b8f.h2.feedac5fb69654cb-cgu.04.rcgu.o:(hashbrown::raw::inner::RawTable$LT$T$C$A$GT$::reserve_rehash::hb74225c1cb92b06f) in archive /checkout/obj/build/ct/iron/target/debug/deps/libh2-7f455a0f03742b8f.rlib
          collect2: error: ld returned 1 exit status
          

error: could not compile `mount` (example "double_mount") due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: linking with `cc` failed: exit status: 1
  |
  = note:  "cc" "-m64" "/tmp/rustcnHGdH5/symbols.o" "<202 object files omitted>" "-Wl,--as-needed" "-Wl,-Bstatic" "/checkout/obj/build/ct/iron/target/debug/deps/{libmount-6e9c66b312103458.rlib,libsequence_trie-5a35fa00d54735d5.rlib,libiron-fc0d0ae2b658f0c7.rlib,libmodifier-4fb1f7e18ad72722.rlib,liburl-2215c3d75c1b8584.rlib,libpercent_encoding-c749af53d9e74952.rlib,libidna-e7e6c70386fa9498.rlib,libunicode_normalization-7e7d0af3110d8833.rlib,libtinyvec-ff29d6288924b12a.rlib,libtinyvec_macros-9107a8cc40680900.rlib,libunicode_bidi-696686a2806343bd.rlib,libmatches-46d1528f77d431eb.rlib,libplugin-f3d0c71b95e43ed1.rlib,libtypemap-093486427740f16d.rlib,libunsafe_any-4930bbaaf7015225.rlib,libtraitobject-340c08327ba62746.rlib,libmime_guess-69adc7798ff3a188.rlib,libunicase-4060a32050b12bb2.rlib,libmime-331a3ecb015ebe30.rlib,libhyper-4cb0a79202e50e61.rlib,libwant-6f13659271f508f4.rlib,libtry_lock-3d6719b540eeb8b5.rlib,libtokio_tcp-d5b94f13d1b96cb6.rlib,libtokio-de0d43b6b561a441.rlib,libtokio_timer-77f3ccedc641dfef.rlib,libtokio_threadpool-6ed6059afa757ecf.rlib,libcrossbeam_queue-18a1a169f95c2544.rlib,libcrossbeam_deque-52bc3a222f80faea.rlib,libcrossbeam_epoch-0b3cda5c0f51574c.rlib,libmemoffset-8dc40f9b9d911ae1.rlib,libtokio_reactor-42930f800d526309.rlib,libtokio_sync-4ad007f9322fa3e5.rlib,libparking_lot-217325a0d93a6aef.rlib,libparking_lot_core-538ce51d58966a81.rlib,libsmallvec-646270aae7e48443.rlib,libmaybe_uninit-dcfcb0c426b7fe3f.rlib,liblock_api-b71f254baa560a48.rlib,libscopeguard-fb83805fc30b1764.rlib,libtokio_current_thread-f8b4e415ea556471.rlib,libtokio_executor-0dbedb92b823b9f2.rlib,libcrossbeam_utils-8d5f9c2a22510913.rlib,liblazy_static-685cca6a61de9387.rlib,libmio-13e00c0d505d782f.rlib,libtime-fddd10bab50ff5b2.rlib,libnet2-fb030a73ce79c7da.rlib,libcfg_if-399308ddf934bdc9.rlib,libhttparse-96ab9397b6935093.rlib,libhttp_body-c9cb5fb9994bb288.rlib,libtokio_buf-df7e18f9dd7750e0.rlib,libh2-7f455a0f03742b8f.rlib,libindexmap-821677bb675840bc.rlib,libhashbrown-6ea7667305d8e2be.rlib,libstring-4b7f892e145a84c8.rlib,libslab-33d2b2db3b3a1c98.rlib,libtokio_io-d3ec67a4e3810a84.rlib,libhttp-bba1012344b91bed.rlib,libitoa-1c0dfcba93f6c2bd.rlib,libfnv-01d75fecb65da78b.rlib,libbytes-62948dbb652bab20.rlib,libeither-f888497e747bd327.rlib,libiovec-4a836b323fd99988.rlib,libbyteorder-df09cad6ecc0a915.rlib,libfutures_cpupool-cb467489621a9287.rlib,libnum_cpus-123ade2ff149fee0.rlib,liblibc-c1672baac6be931e.rlib,libfutures-674fdf4396be79cd.rlib,liblog-3c76e791db33a3b7.rlib}.rlib" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib/{libstd-*,libpanic_unwind-*,libobject-*,libmemchr-*,libaddr2line-*,libgimli-*,librustc_demangle-*,libstd_detect-*,libhashbrown-*,librustc_std_workspace_alloc-*,libminiz_oxide-*,libadler2-*,libunwind-*,libcfg_if-*,liblibc-*,librustc_std_workspace_core-*,liballoc-*,libcore-*,libcompiler_builtins-*}.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustcnHGdH5/raw-dylibs" "-B<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-fuse-ld=lld" "-Wl,-znostart-stop-gc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/ct/iron/target/debug/examples/mount-2f118f0d4e7adace" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: rust-lld: error: undefined symbol: core::ptr::drop_in_place$LT$usize$GT$::h55c321e72c91cfba
          >>> referenced by mod.rs:686 (/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hashbrown-0.12.3/src/raw/mod.rs:686)
          >>>               h2-7f455a0f03742b8f.h2.feedac5fb69654cb-cgu.04.rcgu.o:(hashbrown::raw::inner::RawTable$LT$T$C$A$GT$::reserve_rehash::hb74225c1cb92b06f) in archive /checkout/obj/build/ct/iron/target/debug/deps/libh2-7f455a0f03742b8f.rlib
          collect2: error: ld returned 1 exit status
          

error: could not compile `mount` (example "mount") due to 1 previous error
error: linking with `cc` failed: exit status: 1
  |
  = note:  "cc" "-m64" "/tmp/rustcGz4YEx/symbols.o" "<199 object files omitted>" "-Wl,--as-needed" "-Wl,-Bstatic" "/checkout/obj/build/ct/iron/target/debug/deps/{libpersistent-7e6acb3741de2fc8.rlib,libiron-fc0d0ae2b658f0c7.rlib,libmodifier-4fb1f7e18ad72722.rlib,liburl-2215c3d75c1b8584.rlib,libpercent_encoding-c749af53d9e74952.rlib,libidna-e7e6c70386fa9498.rlib,libunicode_normalization-7e7d0af3110d8833.rlib,libtinyvec-ff29d6288924b12a.rlib,libtinyvec_macros-9107a8cc40680900.rlib,libunicode_bidi-696686a2806343bd.rlib,libmatches-46d1528f77d431eb.rlib,libplugin-f3d0c71b95e43ed1.rlib,libtypemap-093486427740f16d.rlib,libunsafe_any-4930bbaaf7015225.rlib,libtraitobject-340c08327ba62746.rlib,libmime_guess-69adc7798ff3a188.rlib,libunicase-4060a32050b12bb2.rlib,libmime-331a3ecb015ebe30.rlib,libhyper-4cb0a79202e50e61.rlib,libwant-6f13659271f508f4.rlib,libtry_lock-3d6719b540eeb8b5.rlib,libtokio_tcp-d5b94f13d1b96cb6.rlib,libtokio-de0d43b6b561a441.rlib,libtokio_timer-77f3ccedc641dfef.rlib,libtokio_threadpool-6ed6059afa757ecf.rlib,libcrossbeam_queue-18a1a169f95c2544.rlib,libcrossbeam_deque-52bc3a222f80faea.rlib,libcrossbeam_epoch-0b3cda5c0f51574c.rlib,libmemoffset-8dc40f9b9d911ae1.rlib,libtokio_reactor-42930f800d526309.rlib,libtokio_sync-4ad007f9322fa3e5.rlib,libparking_lot-217325a0d93a6aef.rlib,libparking_lot_core-538ce51d58966a81.rlib,libsmallvec-646270aae7e48443.rlib,libmaybe_uninit-dcfcb0c426b7fe3f.rlib,liblock_api-b71f254baa560a48.rlib,libscopeguard-fb83805fc30b1764.rlib,libtokio_current_thread-f8b4e415ea556471.rlib,libtokio_executor-0dbedb92b823b9f2.rlib,libcrossbeam_utils-8d5f9c2a22510913.rlib,liblazy_static-685cca6a61de9387.rlib,libmio-13e00c0d505d782f.rlib,libtime-fddd10bab50ff5b2.rlib,libnet2-fb030a73ce79c7da.rlib,libcfg_if-399308ddf934bdc9.rlib,libhttparse-96ab9397b6935093.rlib,libhttp_body-c9cb5fb9994bb288.rlib,libtokio_buf-df7e18f9dd7750e0.rlib,libh2-7f455a0f03742b8f.rlib,libindexmap-821677bb675840bc.rlib,libhashbrown-6ea7667305d8e2be.rlib,libstring-4b7f892e145a84c8.rlib,libslab-33d2b2db3b3a1c98.rlib,libtokio_io-d3ec67a4e3810a84.rlib,libhttp-bba1012344b91bed.rlib,libitoa-1c0dfcba93f6c2bd.rlib,libfnv-01d75fecb65da78b.rlib,libbytes-62948dbb652bab20.rlib,libeither-f888497e747bd327.rlib,libiovec-4a836b323fd99988.rlib,libbyteorder-df09cad6ecc0a915.rlib,libfutures_cpupool-cb467489621a9287.rlib,libnum_cpus-123ade2ff149fee0.rlib,liblibc-c1672baac6be931e.rlib,libfutures-674fdf4396be79cd.rlib,liblog-3c76e791db33a3b7.rlib}.rlib" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib/{libstd-*,libpanic_unwind-*,libobject-*,libmemchr-*,libaddr2line-*,libgimli-*,librustc_demangle-*,libstd_detect-*,libhashbrown-*,librustc_std_workspace_alloc-*,libminiz_oxide-*,libadler2-*,libunwind-*,libcfg_if-*,liblibc-*,librustc_std_workspace_core-*,liballoc-*,libcore-*,libcompiler_builtins-*}.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustcGz4YEx/raw-dylibs" "-B<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-fuse-ld=lld" "-Wl,-znostart-stop-gc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/ct/iron/target/debug/examples/hitcounter-770e28d72daf7279" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: rust-lld: error: undefined symbol: core::ptr::drop_in_place$LT$usize$GT$::h55c321e72c91cfba
          >>> referenced by mod.rs:686 (/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hashbrown-0.12.3/src/raw/mod.rs:686)
          >>>               h2-7f455a0f03742b8f.h2.feedac5fb69654cb-cgu.04.rcgu.o:(hashbrown::raw::inner::RawTable$LT$T$C$A$GT$::reserve_rehash::hb74225c1cb92b06f) in archive /checkout/obj/build/ct/iron/target/debug/deps/libh2-7f455a0f03742b8f.rlib
          collect2: error: ld returned 1 exit status
          

error: could not compile `persistent` (example "hitcounter") due to 1 previous error
error: linking with `cc` failed: exit status: 1
  |
  = note:  "cc" "-m64" "/tmp/rustchoWYn4/symbols.o" "<200 object files omitted>" "-Wl,--as-needed" "-Wl,-Bstatic" "/checkout/obj/build/ct/iron/target/debug/deps/{libpersistent-7e6acb3741de2fc8.rlib,libiron-fc0d0ae2b658f0c7.rlib,libmodifier-4fb1f7e18ad72722.rlib,liburl-2215c3d75c1b8584.rlib,libpercent_encoding-c749af53d9e74952.rlib,libidna-e7e6c70386fa9498.rlib,libunicode_normalization-7e7d0af3110d8833.rlib,libtinyvec-ff29d6288924b12a.rlib,libtinyvec_macros-9107a8cc40680900.rlib,libunicode_bidi-696686a2806343bd.rlib,libmatches-46d1528f77d431eb.rlib,libplugin-f3d0c71b95e43ed1.rlib,libtypemap-093486427740f16d.rlib,libunsafe_any-4930bbaaf7015225.rlib,libtraitobject-340c08327ba62746.rlib,libmime_guess-69adc7798ff3a188.rlib,libunicase-4060a32050b12bb2.rlib,libmime-331a3ecb015ebe30.rlib,libhyper-4cb0a79202e50e61.rlib,libwant-6f13659271f508f4.rlib,libtry_lock-3d6719b540eeb8b5.rlib,libtokio_tcp-d5b94f13d1b96cb6.rlib,libtokio-de0d43b6b561a441.rlib,libtokio_timer-77f3ccedc641dfef.rlib,libtokio_threadpool-6ed6059afa757ecf.rlib,libcrossbeam_queue-18a1a169f95c2544.rlib,libcrossbeam_deque-52bc3a222f80faea.rlib,libcrossbeam_epoch-0b3cda5c0f51574c.rlib,libmemoffset-8dc40f9b9d911ae1.rlib,libtokio_reactor-42930f800d526309.rlib,libtokio_sync-4ad007f9322fa3e5.rlib,libparking_lot-217325a0d93a6aef.rlib,libparking_lot_core-538ce51d58966a81.rlib,libsmallvec-646270aae7e48443.rlib,libmaybe_uninit-dcfcb0c426b7fe3f.rlib,liblock_api-b71f254baa560a48.rlib,libscopeguard-fb83805fc30b1764.rlib,libtokio_current_thread-f8b4e415ea556471.rlib,libtokio_executor-0dbedb92b823b9f2.rlib,libcrossbeam_utils-8d5f9c2a22510913.rlib,liblazy_static-685cca6a61de9387.rlib,libmio-13e00c0d505d782f.rlib,libtime-fddd10bab50ff5b2.rlib,libnet2-fb030a73ce79c7da.rlib,libcfg_if-399308ddf934bdc9.rlib,libhttparse-96ab9397b6935093.rlib,libhttp_body-c9cb5fb9994bb288.rlib,libtokio_buf-df7e18f9dd7750e0.rlib,libh2-7f455a0f03742b8f.rlib,libindexmap-821677bb675840bc.rlib,libhashbrown-6ea7667305d8e2be.rlib,libstring-4b7f892e145a84c8.rlib,libslab-33d2b2db3b3a1c98.rlib,libtokio_io-d3ec67a4e3810a84.rlib,libhttp-bba1012344b91bed.rlib,libitoa-1c0dfcba93f6c2bd.rlib,libfnv-01d75fecb65da78b.rlib,libbytes-62948dbb652bab20.rlib,libeither-f888497e747bd327.rlib,libiovec-4a836b323fd99988.rlib,libbyteorder-df09cad6ecc0a915.rlib,libfutures_cpupool-cb467489621a9287.rlib,libnum_cpus-123ade2ff149fee0.rlib,liblibc-c1672baac6be931e.rlib,libfutures-674fdf4396be79cd.rlib,liblog-3c76e791db33a3b7.rlib}.rlib" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib/{libstd-*,libpanic_unwind-*,libobject-*,libmemchr-*,libaddr2line-*,libgimli-*,librustc_demangle-*,libstd_detect-*,libhashbrown-*,librustc_std_workspace_alloc-*,libminiz_oxide-*,libadler2-*,libunwind-*,libcfg_if-*,liblibc-*,librustc_std_workspace_core-*,liballoc-*,libcore-*,libcompiler_builtins-*}.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustchoWYn4/raw-dylibs" "-B<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-fuse-ld=lld" "-Wl,-znostart-stop-gc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/ct/iron/target/debug/examples/variable_passing-5782a53e39e15f9f" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: rust-lld: error: undefined symbol: core::ptr::drop_in_place$LT$usize$GT$::h55c321e72c91cfba
          >>> referenced by mod.rs:686 (/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hashbrown-0.12.3/src/raw/mod.rs:686)
          >>>               h2-7f455a0f03742b8f.h2.feedac5fb69654cb-cgu.04.rcgu.o:(hashbrown::raw::inner::RawTable$LT$T$C$A$GT$::reserve_rehash::hb74225c1cb92b06f) in archive /checkout/obj/build/ct/iron/target/debug/deps/libh2-7f455a0f03742b8f.rlib
          collect2: error: ld returned 1 exit status
          

error: could not compile `persistent` (example "variable_passing") due to 1 previous error

thread 'main' panicked at src/tools/cargotest/main.rs:123:9:
tests failed for https://github.com/iron/iron
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:56:25
make: *** [Makefile:51: check-aux] Error 1
  local time: Sat Jun 21 18:52:37 UTC 2025
  network time: Sat, 21 Jun 2025 18:52:37 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
```

</details>

--- a/compiler/rustc_monomorphize/src/collector.rs
+++ b/compiler/rustc_monomorphize/src/collector.rs
@@ -949,6 +949,9 @@ fn visit_instance_use<'tcx>(
         }
         ty::InstanceKind::DropGlue(_, None) => {
             // Don't need to emit noop drop glue if we are calling directly.
+            //
+            // Note that we also optimize away the call to visit_instance_use in vtable construction
+            // (see create_mono_items_for_vtable_methods).
             if !is_direct_call {
                 output.push(create_fn_mono_item(tcx, instance, source));
             }
@@ -1177,8 +1180,13 @@ fn create_mono_items_for_vtable_methods<'tcx>(
         output.extend(methods);
     }
 
-    // Also add the destructor.
-    visit_drop_use(tcx, impl_ty, false, source, output);
+    // Also add the destructor, if it's necessary.
+    //
+    // This matches the check in vtable_allocation_provider in middle/ty/vtable.rs,
+    // if we don't need drop we're not adding an actual pointer to the vtable.
+    if impl_ty.needs_drop(tcx, ty::TypingEnv::fully_monomorphized()) {
+        visit_drop_use(tcx, impl_ty, false, source, output);
+    }
 }
 
 /// Scans the CTFE alloc in order to find function pointers and statics that must be monomorphized.

--- a/tests/codegen-units/item-collection/drop-glue-noop.rs
+++ b/tests/codegen-units/item-collection/drop-glue-noop.rs
@@ -1,0 +1,23 @@
+//@ compile-flags:-Clink-dead-code -Zmir-opt-level=0
+
+#![deny(dead_code)]
+#![crate_type = "lib"]
+
+//~ MONO_ITEM fn start
+#[no_mangle]
+pub fn start(_: isize, _: *const *const u8) -> isize {
+    // No item produced for this, it's a no-op drop and so is removed.
+    unsafe {
+        std::ptr::drop_in_place::<u32>(&mut 0);
+    }
+
+    // No choice but to codegen for indirect drop as a function pointer, since we have to produce a
+    // function with the right signature. In vtables we can avoid that (tested in
+    // instantiation-through-vtable.rs) because we special case null pointer for drop glue since
+    // #122662.
+    //
+    //~ MONO_ITEM fn std::ptr::drop_in_place::<u64> - shim(None) @@ drop_glue_noop-cgu.0[External]
+    std::ptr::drop_in_place::<u64> as unsafe fn(*mut u64);
+
+    0
+}

--- a/tests/codegen-units/item-collection/instantiation-through-vtable.rs
+++ b/tests/codegen-units/item-collection/instantiation-through-vtable.rs
@@ -24,7 +24,6 @@ impl<T> Trait for Struct<T> {
 pub fn start(_: isize, _: *const *const u8) -> isize {
     let s1 = Struct { _a: 0u32 };
 
-    //~ MONO_ITEM fn std::ptr::drop_in_place::<Struct<u32>> - shim(None) @@ instantiation_through_vtable-cgu.0[External]
     //~ MONO_ITEM fn <Struct<u32> as Trait>::foo
     //~ MONO_ITEM fn <Struct<u32> as Trait>::bar
     let r1 = &s1 as &Trait;
@@ -32,7 +31,6 @@ pub fn start(_: isize, _: *const *const u8) -> isize {
     r1.bar();
 
     let s1 = Struct { _a: 0u64 };
-    //~ MONO_ITEM fn std::ptr::drop_in_place::<Struct<u64>> - shim(None) @@ instantiation_through_vtable-cgu.0[External]
     //~ MONO_ITEM fn <Struct<u64> as Trait>::foo
     //~ MONO_ITEM fn <Struct<u64> as Trait>::bar
     let _ = &s1 as &Trait;

--- a/tests/codegen-units/item-collection/non-generic-closures.rs
+++ b/tests/codegen-units/item-collection/non-generic-closures.rs
@@ -1,4 +1,4 @@
-//@ compile-flags:-Clink-dead-code -Zinline-mir=no
+//@ compile-flags:-Clink-dead-code -Zinline-mir=no -O
 
 #![deny(dead_code)]
 #![crate_type = "lib"]
@@ -22,9 +22,8 @@ fn assigned_to_variable_but_not_executed() {
 //~ MONO_ITEM fn assigned_to_variable_executed_indirectly @@ non_generic_closures-cgu.0[External]
 fn assigned_to_variable_executed_indirectly() {
     //~ MONO_ITEM fn assigned_to_variable_executed_indirectly::{closure#0} @@ non_generic_closures-cgu.0[External]
-    //~ MONO_ITEM fn <{closure@TEST_PATH:28:13: 28:21} as std::ops::FnOnce<(i32,)>>::call_once - shim @@ non_generic_closures-cgu.0[External]
-    //~ MONO_ITEM fn <{closure@TEST_PATH:28:13: 28:21} as std::ops::FnOnce<(i32,)>>::call_once - shim(vtable) @@ non_generic_closures-cgu.0[External]
-    //~ MONO_ITEM fn std::ptr::drop_in_place::<{closure@TEST_PATH:28:13: 28:21}> - shim(None) @@ non_generic_closures-cgu.0[External]
+    //~ MONO_ITEM fn <{closure@TEST_PATH:27:13: 27:21} as std::ops::FnOnce<(i32,)>>::call_once - shim @@ non_generic_closures-cgu.0[External]
+    //~ MONO_ITEM fn <{closure@TEST_PATH:27:13: 27:21} as std::ops::FnOnce<(i32,)>>::call_once - shim(vtable) @@ non_generic_closures-cgu.0[External]
     let f = |a: i32| {
         let _ = a + 2;
     };
@@ -40,6 +39,20 @@ fn assigned_to_variable_executed_directly() {
     f(4);
 }
 
+// Make sure we generate mono items for stateful closures that need dropping
+//~ MONO_ITEM fn with_drop @@ non_generic_closures-cgu.0[External]
+fn with_drop(v: PresentDrop) {
+    //~ MONO_ITEM fn with_drop::{closure#0} @@ non_generic_closures-cgu.0[External]
+    //~ MONO_ITEM fn std::ptr::drop_in_place::<PresentDrop> - shim(Some(PresentDrop)) @@ non_generic_closures-cgu.0[Internal]
+    //~ MONO_ITEM fn std::ptr::drop_in_place::<{closure@TEST_PATH:49:14: 49:24}> - shim(Some({closure@TEST_PATH:49:14: 49:24})) @@ non_generic_closures-cgu.0[Internal]
+
+    let _f = |a: usize| {
+        let _ = a + 2;
+        //~ MONO_ITEM fn std::mem::drop::<PresentDrop> @@ non_generic_closures-cgu.0[External]
+        drop(v);
+    };
+}
+
 //~ MONO_ITEM fn start @@ non_generic_closures-cgu.0[External]
 #[no_mangle]
 pub fn start(_: isize, _: *const *const u8) -> isize {
@@ -47,6 +60,7 @@ pub fn start(_: isize, _: *const *const u8) -> isize {
     assigned_to_variable_but_not_executed();
     assigned_to_variable_executed_directly();
     assigned_to_variable_executed_indirectly();
+    with_drop(PresentDrop);
 
     0
 }
@@ -54,4 +68,11 @@ pub fn start(_: isize, _: *const *const u8) -> isize {
 //~ MONO_ITEM fn run_closure @@ non_generic_closures-cgu.0[External]
 fn run_closure(f: &Fn(i32)) {
     f(3);
+}
+
+struct PresentDrop;
+
+impl Drop for PresentDrop {
+    //~ MONO_ITEM fn <PresentDrop as std::ops::Drop>::drop @@ non_generic_closures-cgu.0[External]
+    fn drop(&mut self) {}
 }

--- a/tests/codegen-units/item-collection/unsizing.rs
+++ b/tests/codegen-units/item-collection/unsizing.rs
@@ -1,4 +1,4 @@
-//@ compile-flags:-Zmir-opt-level=0
+//@ compile-flags:-Zmir-opt-level=0 -O
 
 #![deny(dead_code)]
 #![feature(coerce_unsized)]
@@ -42,32 +42,46 @@ struct Wrapper<T: ?Sized>(#[allow(dead_code)] *const T);
 
 impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<Wrapper<U>> for Wrapper<T> {}
 
+struct PresentDrop;
+
+impl Drop for PresentDrop {
+    fn drop(&mut self) {}
+}
+
+// Custom Coercion Case
+impl Trait for PresentDrop {
+    fn foo(&self) {}
+}
+
 //~ MONO_ITEM fn start
 #[no_mangle]
 pub fn start(_: isize, _: *const *const u8) -> isize {
     // simple case
     let bool_sized = &true;
-    //~ MONO_ITEM fn std::ptr::drop_in_place::<bool> - shim(None) @@ unsizing-cgu.0[Internal]
     //~ MONO_ITEM fn <bool as Trait>::foo
     let _bool_unsized = bool_sized as &Trait;
 
     let char_sized = &'a';
 
-    //~ MONO_ITEM fn std::ptr::drop_in_place::<char> - shim(None) @@ unsizing-cgu.0[Internal]
     //~ MONO_ITEM fn <char as Trait>::foo
     let _char_unsized = char_sized as &Trait;
 
     // struct field
     let struct_sized = &Struct { _a: 1, _b: 2, _c: 3.0f64 };
-    //~ MONO_ITEM fn std::ptr::drop_in_place::<f64> - shim(None) @@ unsizing-cgu.0[Internal]
     //~ MONO_ITEM fn <f64 as Trait>::foo
     let _struct_unsized = struct_sized as &Struct<Trait>;
 
     // custom coercion
     let wrapper_sized = Wrapper(&0u32);
-    //~ MONO_ITEM fn std::ptr::drop_in_place::<u32> - shim(None) @@ unsizing-cgu.0[Internal]
     //~ MONO_ITEM fn <u32 as Trait>::foo
     let _wrapper_sized = wrapper_sized as Wrapper<Trait>;
+
+    // with drop
+    let droppable = &PresentDrop;
+    //~ MONO_ITEM fn <PresentDrop as std::ops::Drop>::drop @@ unsizing-cgu.0[Internal]
+    //~ MONO_ITEM fn std::ptr::drop_in_place::<PresentDrop> - shim(Some(PresentDrop)) @@ unsizing-cgu.0[Internal]
+    //~ MONO_ITEM fn <PresentDrop as Trait>::foo
+    let droppable = droppable as &dyn Trait;
 
     false.foo();
 

--- a/tests/codegen/sanitizer/cfi/emit-type-metadata-id-itanium-cxx-abi-drop-in-place.rs
+++ b/tests/codegen/sanitizer/cfi/emit-type-metadata-id-itanium-cxx-abi-drop-in-place.rs
@@ -1,5 +1,9 @@
 // Verifies that type metadata identifiers for drop functions are emitted correctly.
 //
+// Non needs_drop drop glue isn't codegen'd at all, so we don't try to check the IDs there. But we
+// do check it's not emitted which should help catch bugs if we do start generating it again in the
+// future.
+//
 //@ needs-sanitizer-cfi
 //@ compile-flags: -Clto -Cno-prepopulate-passes -Copt-level=0 -Zsanitizer=cfi -Ctarget-feature=-crt-static
 
@@ -10,18 +14,18 @@
 // CHECK:       call i1 @llvm.type.test(ptr {{%.+}}, metadata !"_ZTSFvPu3dynIu{{[0-9]+}}NtNtNtC{{[[:print:]]+}}_4core3ops4drop4Dropu6regionEE")
 
 struct EmptyDrop;
-// CHECK: define{{.*}}4core3ptr{{[0-9]+}}drop_in_place$LT${{.*}}EmptyDrop$GT${{.*}}!type ![[TYPE1]] !type !{{[0-9]+}} !type !{{[0-9]+}} !type !{{[0-9]+}}
+// CHECK-NOT: define{{.*}}4core3ptr{{[0-9]+}}drop_in_place$LT${{.*}}EmptyDrop$GT${{.*}}!type ![[TYPE1]] !type !{{[0-9]+}} !type !{{[0-9]+}} !type !{{[0-9]+}}
 
-struct NonEmptyDrop;
+struct PresentDrop;
 
-impl Drop for NonEmptyDrop {
+impl Drop for PresentDrop {
     fn drop(&mut self) {}
-    // CHECK: define{{.*}}4core3ptr{{[0-9]+}}drop_in_place$LT${{.*}}NonEmptyDrop$GT${{.*}}!type ![[TYPE1]] !type !{{[0-9]+}} !type !{{[0-9]+}} !type !{{[0-9]+}}
+    // CHECK: define{{.*}}4core3ptr{{[0-9]+}}drop_in_place$LT${{.*}}PresentDrop$GT${{.*}}!type ![[TYPE1]] !type !{{[0-9]+}} !type !{{[0-9]+}} !type !{{[0-9]+}}
 }
 
 pub fn foo() {
     let _ = Box::new(EmptyDrop) as Box<dyn Send>;
-    let _ = Box::new(NonEmptyDrop) as Box<dyn Send>;
+    let _ = Box::new(PresentDrop) as Box<dyn Send>;
 }
 
 // CHECK: ![[TYPE1]] = !{i64 0, !"_ZTSFvPu3dynIu{{[0-9]+}}NtNtNtC{{[[:print:]]+}}_4core3ops4drop4Dropu6regionEE"}

